File: samples/csharp_dotnetcore/52.enterprise-bot/ServiceClients/GraphClient.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using Microsoft.Graph;

namespace EnterpriseBot
{
    public class GraphClient
    {
        private readonly string _token;

        public GraphClient(string token)
        {
            _token = token;
        }

        public async Task<User> GetMe()
        {
            var graphClient = GetAuthenticatedClient();
            var me = await graphClient.Me.Request().GetAsync();
            return me;
        }

        private GraphServiceClient GetAuthenticatedClient()
        {
            var graphClient = new GraphServiceClient(
                new DelegateAuthenticationProvider(
                    async (requestMessage) =>
                    {
                        await Task.Run(() =>
                        {
                            var accessToken = _token;

                            // Append the access token to the request.
                            requestMessage.Headers.Authorization = new AuthenticationHeaderValue("bearer", accessToken);

                            // Get event times in the current time zone.
                            requestMessage.Headers.Add("Prefer", "outlook.timezone=\"" + TimeZoneInfo.Local.Id + "\"");
                        });
                    }));

            return graphClient;
        }
    }
}

File: samples/javascript_typescript/52.enterprise-bot/src/serviceClients/graphClient.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { Client, GraphError } from "@microsoft/microsoft-graph-client";
import { User } from "@microsoft/microsoft-graph-types";

export class GraphClient {
    private readonly _token: string;

    constructor(token: string) {
        this._token = token;
    }

    public getMe(): Promise<User> {
        return new Promise((resolve, reject) => {
            const client = this.getAuthenticatedClient();
            client.api("/me").select("displayName")
            .get((err: GraphError, res: User) => {
                if (err) { return reject(err); }
                return resolve(res);
            });
        });
    }

    private getAuthenticatedClient(): Client {
        return Client.init({
            authProvider: (done) => {
                done(null, this._token);
            },
        });
    }
}
